geo/src/algorithm/intersects/mod.rs

```diff
@@ -1,278 +1,114 @@
+// whenever the Rhs is a "simpler geometry" than the target
+// type, and use the macro for the reverse impl.
+#[macro_use]
+macro_rules! symmetric_intersects_impl {
+    ($t:ty, $k:ty, $bounds:tt $(+ $more_bounds:tt )*) => {
+        impl<T> $crate::algorithm::intersects::Intersects<$k> for $t
+        where T: $bounds $(+ $more_bounds)*
+        {
+            fn intersects(&self, rhs: &$k) -> bool {
+                rhs.intersects(self)
             }
-        }
-    }
-}
-
-impl<T> Intersects<Line<T>> for Point<T>
-where
-    T: Float,
-{
-    fn intersects(&self, line: &Line<T>) -> bool {
-        line.intersects(self)
-    }
-}
-
-impl<T> Intersects<Line<T>> for Line<T>
-where
-    T: Float,
-{
-    fn intersects(&self, line: &Line<T>) -> bool {
-        // Using Cramer's Rule:
-        // https://en.wikipedia.org/wiki/Intersection_%28Euclidean_geometry%29#Two_line_segments
-        let a1 = self.dx();
-        let a2 = self.dy();
-        let b1 = -line.dx();
-        let b2 = -line.dy();
-        let c1 = line.start.x - self.start.x;
-        let c2 = line.start.y - self.start.y;
 
-        let d = a1 * b2 - a2 * b1;
-        if d == T::zero() {
-            let (self_start, self_end) = self.points();
-            let (other_start, other_end) = line.points();
-            // lines are parallel
-            // return true iff at least one endpoint intersects the other line
-            self_start.intersects(line)
-                || self_end.intersects(line)
-                || other_start.intersects(self)
-                || other_end.intersects(self)
-        } else {
-            let s = (c1 * b2 - c2 * b1) / d;
-            let t = (a1 * c2 - a2 * c1) / d;
-            (T::zero() <= s) && (s <= T::one()) && (T::zero() <= t) && (t <= T::one())
         }
-    }
+    };
 }
 
-impl<T> Intersects<LineString<T>> for Line<T>
-where
-    T: Float,
-{
-    fn intersects(&self, linestring: &LineString<T>) -> bool {
-        linestring.lines().any(|line| self.intersects(&line))
-    }
-}
+use crate::kernels::HasKernel;
 
-impl<T> Intersects<Line<T>> for LineString<T>
-where
-    T: Float,
-{
-    fn intersects(&self, line: &Line<T>) -> bool {
-        line.intersects(self)
-    }
-}
+symmetric_intersects_impl!(Coordinate<T>, Point<T>, CoordinateType);
+symmetric_intersects_impl!(Coordinate<T>, Line<T>, HasKernel);
 
-impl<T> Intersects<Polygon<T>> for Line<T>
-where
-    T: Float,
-{
-    fn intersects(&self, p: &Polygon<T>) -> bool {
-        p.exterior().intersects(self)
-            || p.interiors().iter().any(|inner| inner.intersects(self))
-            || p.contains(&self.start_point())
-            || p.contains(&self.end_point())
-    }
-}
+symmetric_intersects_impl!(Point<T>, Line<T>, HasKernel);
 
-impl<T> Intersects<Line<T>> for Polygon<T>
-where
-    T: Float,
-{
-    fn intersects(&self, line: &Line<T>) -> bool {
-        line.intersects(self)
-    }
-}
+symmetric_intersects_impl!(Line<T>, LineString<T>, HasKernel);
```

Comment:
Not a big deal, so take it or leave it, but for someone trying to find how the intersection logic is actually implemented, it might be helpful to collocate the macro invocation next to the implementation.

i.e. this one could go next to the `impl<T> Intersects<Line<T>> for LineString<T>`

Another benefit to organizing it that way is it might be easier for the implementor (or reviewer 😉)to audit that each `impl` is preceded by the macro invocation than it is to cross reference that the sum of all appropriate calls exists in this file.

Reply:
Good point; will do.